gcc/config/riscv/crypto.md

```diff
@@ -0,0 +1,302 @@
+    (unspec:SI
+      [(match_operand:SI 1 "register_operand" "r")
+        (match_operand:SI 2 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && !TARGET_64BIT"
+  "sha512sig1l\t%0,%1,%2")
+
+(define_insn "riscv_sha512sum0r"
+  [(set (match_operand:SI 0 "register_operand" "=r")
+    (unspec:SI
+      [(match_operand:SI 1 "register_operand" "r")
+        (match_operand:SI 2 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && !TARGET_64BIT"
+  "sha512sum0r\t%0,%1,%2")
+
+(define_insn "riscv_sha512sum1r"
+  [(set (match_operand:SI 0 "register_operand" "=r")
+    (unspec:SI
+      [(match_operand:SI 1 "register_operand" "r")
+        (match_operand:SI 2 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && !TARGET_64BIT"
+  "sha512sum1r\t%0,%1,%2")
+
+
+;; Zknh - SHA512 (RV64)
+
+(define_insn "riscv_sha512sig0"
+  [(set (match_operand:DI 0 "register_operand" "=r")
+    (unspec:DI
+      [(match_operand:DI 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && TARGET_64BIT"
+  "sha512sig0\t%0,%1")
+
+(define_insn "riscv_sha512sig1"
+  [(set (match_operand:DI 0 "register_operand" "=r")
+    (unspec:DI
+      [(match_operand:DI 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && TARGET_64BIT"
+  "sha512sig1\t%0,%1")
+
+(define_insn "riscv_sha512sum0"
+  [(set (match_operand:DI 0 "register_operand" "=r")
+    (unspec:DI
+      [(match_operand:DI 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && TARGET_64BIT"
+  "sha512sum0\t%0,%1")
+
+(define_insn "riscv_sha512sum1"
+  [(set (match_operand:DI 0 "register_operand" "=r")
+    (unspec:DI
+      [(match_operand:DI 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKNH && TARGET_64BIT"
+  "sha512sum1\t%0,%1")
+
+
+;; Zksh - SM3
+
+(define_insn "riscv_sm3p0_<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+    (unspec:X
+      [(match_operand:X 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKSH"
+  "sm3p0\t%0,%1")
+
+(define_insn "riscv_sm3p1_<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+    (unspec:X
+      [(match_operand:X 1 "register_operand" "r")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKSH"
+  "sm3p1\t%0,%1")
+
+
+;; Zksed - SM4
+
+(define_insn "riscv_sm4ed_<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+    (unspec:X
+      [(match_operand:X 1 "register_operand" "r")
+        (match_operand:SI 2 "immediate_operand" "")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKSED"
+  "sm4ed\t%0,%1,%2")
+
+(define_insn "riscv_sm4ks_<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+    (unspec:X
+      [(match_operand:X 1 "register_operand" "r")
+        (match_operand:SI 2 "immediate_operand" "")]
+        UNSPEC_CRYPTO))]
+  "TARGET_ZKSED"
+  "sm4ks\t%0,%1,%2")
+
+
+;; Zkr - Entropy Source
+
+(define_insn "riscv_pollentropy_<mode>"
+  [(unspec_volatile [(match_operand:X 0 "register_operand" "=r")] UNSPEC_CRYPTO)]
```

Comment:
I tried to changed this to `(set (match_operand:X 0 "register_operand" "=r") (unspec:X UNSPEC_XXX))`, but it seem that it can't pass the compile, the "unspec" wants a vector after it.

Reply:
Use a dummy `const_int 0` for that, like  `(set (match_operand:X 0 "register_operand" "=r") (unspec:X [(const_int 0)] UNSPEC_XXX))`, this trick are used in several different back-end :)

You can grep by following command to see:
```
grep "\[(const_int 0)\]" * -R |grep unspec
```